src/raftstore/store/local_read.rs

```diff
@@ -0,0 +1,446 @@
+               -> PeerStatus {
+        let tag = format!("[region {}] {}", region.get_id(), peer.get_id());
+        PeerStatus {
+            peer: peer,
+            region: region,
+            tag: tag,
+            term: term,
+            applied_index_term: applied_index_term,
+            leader_id: leader_id,
+            leader_lease_expired_time: None,
+        }
+    }
+
+    pub fn find_peer(&self, peer_id: u64) -> Option<&metapb::Peer> {
+        for peer in self.region.get_peers() {
+            if peer.get_id() == peer_id {
+                return Some(peer);
+            }
+        }
+        None
+    }
+
+    pub fn peer_id(&self) -> u64 {
+        self.peer.get_id()
+    }
+
+    pub fn is_leader(&self) -> bool {
+        self.peer_id() == self.leader_id
+    }
+
+    pub fn should_read_local(&mut self, req: &RaftCmdRequest) -> bool {
+        if (req.has_header() && req.get_header().get_read_quorum()) || !self.is_leader() ||
+           req.get_requests().len() == 0 {
+            return false;
+        }
+
+        // If applied index's term is differ from current raft's term, leader transfer
+        // must happened, if read locally, we may read old value.
+        if self.applied_index_term != self.term {
+            return false;
+        }
+
+        for cmd_req in req.get_requests() {
+            if cmd_req.get_cmd_type() != CmdType::Snap && cmd_req.get_cmd_type() != CmdType::Get {
+                return false;
+            }
+        }
+
+        // If the leader lease has expired, local read should not be performed.
+        if self.leader_lease_expired_time.is_none() {
+            return false;
+        }
+
+        let mut reset_lease_expired_time = false;
+        let mut lost_lease = false;
+        let now = clocktime::raw_now();
+        match self.leader_lease_expired_time.as_ref().unwrap().as_ref() {
+            Either::Left(safe_expired_time) => {
+                if now > *safe_expired_time {
+                    reset_lease_expired_time = true;
+                    lost_lease = true;
+                }
+            }
+            _ => lost_lease = true,
+        }
+        if reset_lease_expired_time {
+            debug!("leader lease expired time {:?} is outdated",
+                   self.leader_lease_expired_time);
+            // Reset leader lease expiring time.
+            self.leader_lease_expired_time = None;
+        }
+        if lost_lease {
+            // Perform a consistent read to Raft quorum and try to renew the leader lease.
+            return false;
+        }
+
+        true
+    }
+}
+
+type Key = Vec<u8>;
+
+pub struct LocalReadHandler {
+    rx: Receiver<Msg>,
+    peers_status: HashMap<u64, PeerStatus>,
+    store_id: u64,
+    engine: Arc<DB>,
+
+    // used when need redirect request to raftstore
+    sendch: SendCh<Msg>,
+    redirect_cmds: HashMap<Uuid, Callback>,
+
+    // region end key -> region id
+    region_ranges: BTreeMap<Key, u64>,
```

Comment:
I think we don't need region_ranges here, if we meet the stale or lease expired error, we can all redirect the request to the raftstore thread.